db/coordination.go

```diff
@@ -5,115 +5,207 @@ import (
 	"database/sql"
 	"time"
 
-	"github.com/jmoiron/sqlx"
+	"github.com/cyverse-de/data-usage-api/config"
+	"github.com/cyverse-de/data-usage-api/util"
 	"github.com/pkg/errors"
 )
 
 type BothDatabases struct {
-	dedb   *DEDatabase
-	icatdb *ICATDatabase
+	deconn        DatabaseTxAccessor
+	icatconn      DatabaseTxAccessor
+	configuration *config.Config
+
+	DERollback func()
+	DECommit   func() error
+
+	ICATRollback func()
+	ICATCommit   func() error
+
+	detx   *DEDatabase
+	icattx *ICATDatabase
 }
 
-func NewBoth(dedb *DEDatabase, icatdb *ICATDatabase) *BothDatabases {
-	return &BothDatabases{dedb: dedb, icatdb: icatdb}
+func NewBoth(dedb DatabaseTxAccessor, icatdb DatabaseTxAccessor, config *config.Config) *BothDatabases {
+	return &BothDatabases{deconn: dedb, icatconn: icatdb, configuration: config}
 }
 
-func NewBothTx(context context.Context, deconn *sqlx.DB, deschema string, icatconn *sqlx.DB, userSuffix, zone string, rootResourceNames []string) (*BothDatabases, func(), func() error, error) {
-	logStats("DE", deconn)
-	detx, err := deconn.BeginTxx(context, nil)
-	if err != nil {
-		return nil, func() {}, func() error { return err }, errors.Wrap(err, "Error creating DE database transaction")
+func (b *BothDatabases) DETx(ctx context.Context) (*DEDatabase, error) {
```

Comment:
This function and ICATTx are both pretty wordy. I'd appreciate if anyone has good ideas on how to shorten them (maybe combining them? not sure)

Reply:
I started actually reviewing the code then got sidetracked on this because I find puzzles like this to be enjoyable. 😆 At any rate, the cleanest way that I've found to shorten these functions that I've found so far is to create a new transaction wrapper type:

``` go
type TxWrapper struct {
	DBName   string
	Tx       *sqlx.Tx
	Rollback func()
	Commit   func() error
}

func NewTxWrapper(ctx context.Context, dbName string, conn DatabaseTxAccessor) (*TxWrapper, error) {
	tx, err := conn.BeginTxx(ctx, nil)
	if err != nil {
		return nil, errors.Wrapf(err, "Error creating %s transaction", dbName)
	}

	wrapper := &TxWrapper{
		DBName: dbName,
		Tx:     tx,
	}
	wrapper.InitRollbackFunction()
	wrapper.InitCommitFunction()

	return wrapper, nil
}

func (w *TxWrapper) InitRollbackFunction() {
	w.Rollback = func() {
		alreadyDoneErr := "sql: transaction has already been committed or rolled back"
		err := w.Tx.Rollback()
		if err != nil && err.Error() != alreadyDoneErr {
			e := errors.Wrapf(err, "Error rolling back %s database transaction", w.DBName)
			log.Error(e)
		}
		w.Tx = nil
		w.Rollback = func() {}
		w.Commit = func() error { return nil }
	}
}

func (w *TxWrapper) InitCommitFunction() {
	w.Commit = func() error {
		err := w.Tx.Commit()
		w.Tx = nil
		w.Rollback = func() {}
		w.Commit = func() error { return nil }
		if err != nil {
			e := errors.Wrapf(err, "Error committing %s database transaction", w.DBName)
			log.Error(e)
			return e
		}
		return nil
	}
}
```

With this in place, the `BothDatabases` struct can be shortened to this:

``` go
type BothDatabases struct {
	deconn        DatabaseTxAccessor
	icatconn      DatabaseTxAccessor
	configuration *config.Config

	detxWrapper   *TxWrapper
	icattxWrapper *TxWrapper

	detx   *DEDatabase
	icattx *ICATDatabase
}
```

And the transaction functions can be shortened to this:

``` go
func (b *BothDatabases) DETx(ctx context.Context) (*DEDatabase, error) {
	logStats("DE", b.deconn)
	if b.detx != nil {
		return b.detx, nil
	}

	detxWrapper, err := NewTxWrapper(ctx, "DE", b.deconn)
	if err != nil {
		return nil, errors.Wrap(err, "Error creating DE transaction")
	}

	b.detxWrapper = detxWrapper
	b.detx = NewDE(detxWrapper.Tx, b.configuration)

	return b.detx, nil
}

func (b *BothDatabases) ICATTx(ctx context.Context) (*ICATDatabase, error) {
	logStats("ICAT", b.icatconn)
	if b.icattx != nil {
		return b.icattx, nil
	}

	icattxWrapper, err := NewTxWrapper(ctx, "ICAT", b.icatconn)
	if err != nil {
		return nil, errors.Wrap(err, "Error creating DE transaction")
	}

	b.icattxWrapper = icattxWrapper
	b.icattx = NewICAT(icattxWrapper.Tx, b.configuration)

	return b.icattx, nil
}
```

The drawback is that we then have to define the rollback and commit functions like this:

```
func (b *BothDatabases) DERollback() {
	if b.detxWrapper != nil {
		b.detxWrapper.Rollback()
		b.detxWrapper = nil
		b.detx = nil
	}
}

func (b *BothDatabases) DECommit() error {
	var err error
	if b.detxWrapper != nil {
		err = b.detxWrapper.Commit()
		b.detxWrapper = nil
		b.detx = nil
	}
	return err
}

func (b *BothDatabases) ICATRollback() {
	if b.icattxWrapper != nil {
		b.icattxWrapper.Rollback()
		b.icattxWrapper = nil
		b.icattx = nil
	}
}

func (b *BothDatabases) ICATCommit() error {
	var err error
	if b.icattxWrapper != nil {
		err = b.icattxWrapper.Commit()
		b.icattxWrapper = nil
		b.icattx = nil
	}
	return err
}
```

This does compile, but the usual caveat that this hasn't been tested yet applies here. I also haven't decided whether or not this is actually clearer than the code that you already have. At least this was a fun exercise though. 😆

Reply:
Haha, fair enough. I think I'll keep it how it is for now since yeah, it's a little verbose the changed way too, haha (more of a "more functions vs. long functions" tradeoff, feels like). I wonder if there's a way to roll some of this logic into other types (maybe DatabaseAccessor? idk), but seems like one of those "let it simmer in the back of my head for a month or two" problems at this point.